src/material/badge/badge.ts

```diff
@@ -145,70 +145,47 @@ export class MatBadge extends _MatBadgeBase implements OnDestroy, OnChanges, Can
     return this.position.indexOf('before') === -1;
   }
 
-  ngOnChanges(changes: SimpleChanges) {
-    const contentChange = changes['content'];
-
-    if (contentChange) {
-      const value = contentChange.currentValue;
-      this._hasContent = value != null && `${value}`.trim().length > 0;
-      this._updateTextContent();
-    }
-  }
-
-  ngOnDestroy() {
-    const badgeElement = this._badgeElement;
-
-    if (badgeElement) {
-      if (this.description) {
-        this._ariaDescriber.removeDescription(badgeElement, this.description);
-      }
-
-      // When creating a badge through the Renderer, Angular will keep it in an index.
-      // We have to destroy it ourselves, otherwise it'll be retained in memory.
-      if (this._renderer.destroyNode) {
-        this._renderer.destroyNode(badgeElement);
-      }
-    }
-  }
-
   /**
-   * Gets the element into which the badge's content is being rendered.
-   * Undefined if the element hasn't been created (e.g. if the badge doesn't have content).
+   * Gets the element into which the badge's content is being rendered. Undefined if the element
+   * hasn't been created (e.g. if the badge doesn't have content).
    */
   getBadgeElement(): HTMLElement | undefined {
     return this._badgeElement;
   }
 
-  /** Injects a span element into the DOM with the content. */
-  private _updateTextContent(): HTMLSpanElement {
-    if (!this._badgeElement) {
-      this._badgeElement = this._createBadgeElement();
-    } else {
-      this._badgeElement.textContent = this._stringifyContent();
+  ngOnInit() {
+    // We may have server-side rendered badge that we need to clear.
+    // We need to do this in ngOnInit because the full content of the component
+    // on which the badge is attached won't necessarily be in the DOM until this point.
+    this._clearExistingBadges();
+  }
+
+  ngOnDestroy() {
+    // When creating a badge through the Renderer, Angular will keep it in an index.
+    // We have to destroy it ourselves, otherwise it'll be retained in memory.
```

Comment:
Worth noting that this is a ViewEngine-specific problem.

Reply:
Updated comment to note that